### ISSUE TITLE
Redesign progressbar interface

### ISSUE THREAD

**Issue by fushinari** [goodmami/wn#87]
https://github.com/goodmami/wn/blob/180a84d526b399151637cdee87513b40d905687d/wn/_download.py#L83

Here, the `max` parameter overrides the built-in function `max`. Maybe renaming it should be considered?

**Comment by goodmami** [comment]
It's not a reserved keyword like `in`, `def`, or `True`, so it's actually ok. And it's not uncommon to use these when it's a descriptive variable name and the original object/function is not used (also `file` (in Python 2), `type`, and even `set` or `list`, although I don't reuse these latter two). Nevertheless I can see how it can be confusing, and some linters (like pylint, I think) will warn about reusing these names. The [progress](https://github.com/verigak/progress/) library also uses `max`, but [tqdm](https://github.com/tqdm/tqdm/) uses `total`, which isn't so bad.

Relatedly, I was thinking of making changes to the way callbacks are handled so it's more compatible with these other progress bar tools. So far I think you're the only one customizing them so I'd like to make sure you're ok with the changes. Basically, I want to separate the main update function so it only takes the single increment value and all other updates (status, changes to count or max, etc.) are done in a separate function (e.g., `bar.set(status='Initializing')`), but it is only called if the method exists.

**Comment by fushinari** [comment]
I see, then I guess it's fine.

Any breaking changes that I can adapt my code to is okay for me. :)

I already freeze wn's version now in the Flatpak version (which will be the supported version from now on), so any breaking changes are fine as long as I can adapt to them later on.

**Comment by goodmami** [comment]
Ok I'm repurposing this one for the new design of the progress bar. While I'm at it I can rename `max` to `total`. Here's what I'm thinking:

A base class `wn.util.ProgressHandler`:

```python
class ProgressHandler:

    def __init__(
        self,
        *,
        message: str = '',
        count: int = 0,
        total: int = 0,
        unit: str = '',
        status: str = '',
        file: TextIO = sys.stderr,
    ):
        self.file = file
        self.kwargs = {
            'count': count,
            'total': total,
            'message': message,
            'unit': unit,
            'status': status,
        }

    def update(self, n: int = 1) -> None:
        self.kwargs['count'] += n

    def set(self, **kwargs) -> None:
        self.kwargs.update(**kwargs)
        self.update(0)

    def flash(self, message: str) -> None:
        pass

    def close(self, message: str = None) -> None:
        pass
```

And a derived class `wn.util.ProgressBar`, which is the default for `wn.download()`, `wn.add()`, etc. That is, the `progress_handler` parameter of `wn.download()` and `wn.add()` should be a `ProgressHandler` subclass which will be instantiated accordingly. It can also be `None`, which disables the progress bar. Examples of use:

```python
wn.download(..., progress_handler=None)  # disable
wn.download(..., progress_handler=wn.util.ProgressBar)  # new instance will be created

class MyHandler(wn.util.ProgressHandler):
    def update(self, n: int = 1) -> None:
        do_something(n)

wn.download(..., progress_handler=MyHandler)  # custom class
```

**Comment by goodmami** [comment]
@fushinari note that the `flash()` method might be useful for you. It is used by `wn.add()` (and thus `wn.download()`) for messages about checking and reading resource files. E.g.:

```python
        ...
        # abort if any lexicon in *source* is already added
        progress.flash(f'Checking {source!s}')
        all_infos = list(_precheck(source, cur))

        if not all_infos:
            progress.flash(f'{source}: No lexicons found')
            return
        elif all(info.get('skip', False) for info in all_infos):
            progress.flash(f'{source}: Some or all lexicons already added')
            return

        # all clear, try to add them
        progress.flash(f'Reading {source!s}')
        ...
```

Currently they are written straight to stderr, but with the proposed change they will be handled by the ProgressHandler object as shown above, so you could display or discard the messages as you wish.

**Comment by fushinari** [comment]
LGTM! I don't know how exactly I'd make use of `flash()` right now but it does seem useful and I'm looking forward to trying it out.

**Comment by goodmami** [comment]
Great, thanks for the feedback. I'm using `flash()` in the `ProgressBar` class for messages that clear the progress bar and print on the whole line. It seems a lot cleaner than my previous solution. At a minimum, you can just do nothing with the `flash()` calls and not worry about the stderr buffer filling up with those messages.

**Comment by goodmami** [comment]
I released v0.5.0 with these changes. I pushed back some database changes to the next release, so you should be able to try this out without rebuilding the database. See https://wn.readthedocs.io/en/latest/api/wn.util.html

**Comment by fushinari** [comment]
Thank you! I'll be checking it out tomorrow.
